6_nodejs/week2/app.js

```diff
@@ -0,0 +1,87 @@
+import express from "express";
+const app = express();
+import fs from "fs";
+const data = fs.readFileSync("./documents.json");
+let documents = JSON.parse(data);
+const port = process.env.PORT || 3000;
+
+// Support parsing JSON requests
+app.use(express.json());
+
+app.get("/", (req, res) => {
+  res.send("This is a search engine");
+});
+//part1
+app.get("/search", (req, res) => {
+  const query = req.query.q; // Extract the 'q' query parameter from the request URL
+  if (!query) {
+    res.json(documents); // If 'q' is not provided, it returns all documents
+  } else {
+    // filter documents' objects, as doc, based on whether any field matches the query string
+    const filteredDocuments = documents.filter((doc) =>
+      Object.values(doc).some(
+        (value) => typeof value === "string" && value.includes(query)
+      )
+    );
+    if (filteredDocuments) {
+      res.json(filteredDocuments);
+    } else {
+      res.send("no match found");
```

Comment:
I think in this case it would be better if the endpoint returns an empty array instead of a string. Imagine receiving this data in the client: one time you get an array of objects, the other time you get a string and you have two handle both of those cases. It would be way earlier just to get an empty array if there are no search results found.